extensions/wikia/BannerNotifications/js/BannerNotifications.js

```diff
@@ -0,0 +1,251 @@
+	 * @param {String} content
+	 * @param {String} [type]
+	 * @returns {jQuery}
+	 */
+	function createMarkup(content, type) {
+		return $(
+			mustache.render(templates.BannerNotifications, {
+				imageSource: closeImageSource,
+				content: content
+			})
+		).addClass(type).hide();
+	}
+
+	/**
+	 * Adds given markup to DOM
+	 * @param {jQuery} $element
+	 * @param {jQuery} $parentElement
+	 */
+	function addToDOM($element, $parentElement) {
+		// allow notification wrapper element to be passed by extension
+		var $parent = $parentElement || isModalShown() ? modal : $pageContainer,
+			$bannerNotificationsWrapper = $parent.find('.banner-notifications-wrapper');
+		if (!$bannerNotificationsWrapper.length) {
+			$bannerNotificationsWrapper = $('<div></div>').addClass('banner-notifications-wrapper');
+			$parent.prepend($bannerNotificationsWrapper);
+		}
+		$bannerNotificationsWrapper.prepend($element);
+
+		$element.fadeIn('slow');
+	}
+
+	function BannerNotification(content, type, $parent, timeout) {
+		if (content instanceof jQuery && content.hasClass('banner-notification')) {
+			//create a notification object from already existing markup
+			this.content = content.find('.msg').html();
+			this.$element = content;
+			this.$parent = content.parent();
+			this.hidden = false;
+		} else {
+			this.content = content;
+			this.$element = null;
+			this.$parent = $parent;
+			this.hidden = true;
+			this.type = type;
+			this.timeout = timeout;
+		}
+	}
+
+	BannerNotification.prototype.show = function () {
+		var self;
+		if (!this.hidden) {
+			return this;
+		}
+		if (!this.$element) {
+			this.$element = createMarkup(this.content, this.type);
+		}
+		this.setType(this.type);
+
+		setUpClose(this);
+
+		// Modal notifications have no close button so set a timeout
+		if (isModalShown() && typeof this.timeout !== 'number') {
+			this.timeout = defaultTimeout;
+		}
+
+		addToDOM(this.$element, this.$parent);
+
+		this.hidden = false;
+
+		// Close notification after specified amount of time
+		if (typeof this.timeout === 'number') {
+			self = this;
+			setTimeout(function () {
+				self.hide();
+			}, this.timeout);
+		}
+		return this;
+	};
+
+	BannerNotification.prototype.hide = function (callback) {
+		if (!this.hidden) {
+			removeFromDOM(this.$element, callback);
+			this.$element = null;
+			this.hidden = true;
+		}
+		return this;
+	};
+
+	BannerNotification.prototype.setType = function (type) {
+		if (type !== this.type && types.hasOwnProperty(type)) {
+			this.type = type;
+			if (this.$element) {
+				this.$element
+					.removeClass(classes)
+					.addClass(type);
+			}
+		}
+		return this;
+	};
+
+	BannerNotification.prototype.setContent = function (content) {
+		if (content && content !== this.content) {
+			this.content = content;
+			if (this.$element) {
+				this.$element
+					.find('.msg')
+					.html(content);
+			}
+		}
+		return this;
+	};
+
+	BannerNotification.prototype.showConnectionError = function () {
```

Comment:
Let's add more flushed out doc blocks for the file. Also, it would be nice to order the functions a bit better too. For example, you could put the private functions below all the prototype methods, and place each method in the order in which it might be called.